feat: Image search example

Green-Sky opened the pull request:
I used the [usearch](https://github.com/unum-cloud/usearch) vector database library, to build a image embedding database and then to search in it by similarity using a text string.

build the database:
```
$ bin/image-search-build -m ../models/ggml-model-f16.bin ../tests/
clip_model_load: loading model from '../models/ggml-model-f16.bin' - please wait....................................................clip_model_load: model size =   288.93 MB / num tensors = 397
clip_model_load: model loadded
main: starting base dir scan of '../tests/'
main: found image file '../tests/white.jpg'
main: found image file '../tests/red_apple.jpg'
```
it generates an `images.usearch` and `images.path` file.

search by similarity to string:
```
$ bin/image-search -m ../models/ggml-model-f16.bin "a red apple"
clip_model_load: loading model from '../models/ggml-model-f16.bin' - please wait....................................................clip_model_load: model size =   288.93 MB / num tensors = 397
clip_model_load: model loadded
search results:
similarity path
  0.655244 /home/green/workspace/clip.cpp/tests/red_apple.jpg
  0.773273 /home/green/workspace/clip.cpp/tests/white.jpg
```

It is a very rough implementation. It requires `c++17` and `std::filesystem` for iterating the filesystem.

A similarity search via an image would be nice too.

basically we could copy all the features from here: https://github.com/yurijmikhalevich/rclip

### TODO:
 - [x] add README.md to example
 - [x] make it compile conditionally (bc c++17 and deps)
 - [ ] solve the usearch memory leak (tracking https://github.com/unum-cloud/usearch/issues/138)
 - [x] tabs to spaces
 - [x] use fetch content instead of committing deps

monatis commented on examples/image-search/usearch/index.hpp:
Instead of copying this file into the repo, you can configure the example's `CMakeLists.txt` to automatically download it. Citing from USearch:

> Alternatively fetch it with CMake:

```cmake

FetchContent_Declare(usearch GIT_REPOSITORY https://github.com/unum-cloud/usearch.git)
FetchContent_MakeAvailable(usearch)
```

And of course, git-ignore it to not accidentally push.

Green-Sky replied:
oh. that actually makes sense too. I forgot that this is at compile/configure time. when I looked at it my brain just assumed it meant we have to include the whole repo.
technically we only want that 1 file. and that usearch repo is kinda full of stuff, but will look into this. I think the cmake version needs to be increased for this too.

monatis replied:
Good concern, but if my memory serves me right fetchContent was introduced in CMake 3.11, and we are setting the minimum required version to 3.12 already.

monatis commented on examples/image-search/build.cpp:
```diff
@@ -0,0 +1,155 @@
+		fprintf(stdout, "%s: starting base dir scan of '%s'\n", __func__, base_dir.c_str());
+
+		for (auto const& dir_entry : std::filesystem::recursive_directory_iterator(base_dir)) {
+			if (!dir_entry.is_regular_file()) {
```
Do we need C++17 just for this? According to their header, USearch requires only C++11. And implementing this recursive directory walk should not be that hard in C++11 as well.

Green-Sky replied:
c++17 filesystem was the easy route. with c++11 we have to use platform specific code. so different code paths for windows/linux etc.

monatis replied:
I'll look into this --if there's no gotcha other than different routes for Windows and unix, I can PR to this branch. I'd like just to support as many people as possible 😀

Green-Sky replied:
I am not targeting merging WITH c++17, but the example disabled by default in cmake.
You(or me) can then patch it later to not use `std::filesystem`.

monatis replied:
Good idea! We can invest more in this later.

Feel free to mark it as ready for review once you're satisfied.

monatis commented on examples/image-search/build.cpp:
```diff
@@ -0,0 +1,155 @@
+			clip_image_preprocess(&img0, &img_res);
+
+			std::vector<float> vec(vec_dim);
+			if (!clip_image_encode(clip_ctx, params.n_threads, img_res, vec.data())) {
```
Why not using a float array directly? We are working with fixed lengths at the end.

Green-Sky replied:
are we? I thought it depends on the specific model used.
but yea, i guess we could reuse the same vector, since it does not change during execution.

monatis replied:
Yeah, it's more like a coding style issue. both this repo and the USearch repo use float arrays instead of `std::vector<float>` to represent embedding vectors.

Green-Sky replied:
> float arrays

but the size is not known at compile time. runtime size declaration is an extension, and not proper c++

monatis replied:
I just meant, for example, copying from `examples/main.cpp`

```C++
    float img_vec[vec_dim];
    if (!clip_image_encode(ctx, params.n_threads, img_res, img_vec))
    {
        return 1;
    }
```

But ok, it was just a nit, and not a strong opinion :).

monatis reviewed (changes requested):
I like the idea to provide an image search example with tiny vector search implementation directly in C++, but I highlighted some points.

Also, let's not use tab characters for indents please.

>  general code clean up. but the rest of the repo is still in a similar state 😝

Things may be messy at early stages to quickly introduce new features, but I'm trying to tidy up the code and README as we make progress in the project. You can always suggest better structures for the code and the overall project.

Green-Sky commented:
> Also, let's not use tab characters for indents please.

Ohhhhh yea. sorry. for got to reverse this :laughing:

Green-Sky commented:
> Things may be messy at early stages to quickly introduce new features, but I'm trying to tidy up the code and README as we make progress in the project. You can always suggest better structures for the code and the overall project.

yea, I also intentionally maked this PR as draft for those reasons. :)

monatis commented:
> yea, I also intentionally maked this PR as draft for those reasons. :)

Great, thank you. Your work really appreciated. And to be honest, I was not aware of Usearch, but instead considering the original hnswlib library for the search example. But Usearch looks like a better candidate.

Green-Sky commented:
forcepushed with `FetchContent` and spaces instead of tabs.

monatis commented on examples/image-search/search.cpp:
```diff
@@ -0,0 +1,115 @@
+    clip_text_encode(clip_ctx, params.n_threads, tokens, txt_vec.data());
+
+    auto results = embd_index.search({txt_vec.data(), txt_vec.size()}, 5);
+
```
Maybe also getting the number of results from CLI arg too?

Green-Sky replied:
yes. i could swear i left a TODO somewhere in the code...

monatis commented:
Do you need an extra hand in this?

Green-Sky commented:
Was doing something completely different last week :)
Will try to get this into a mergeable state this week.

Green-Sky commented:
just rebased and changed some stuff you committed on main

issues that I found:
1. you added cmake options for conditionally disabling tests and examples, i negated them and initialized them to standalone (more convenient)
2. you require c++20 (!) should only be 11 i think. (and 17 for fs in image-search until its changed to something else)
3. you set gcc/clang specific warning and `O3` optimization level indiscriminately https://github.com/monatis/clip.cpp/blob/1f140e027a3b9165a1edcdd4948bb44c3c2eacff/CMakeLists.txt#L136 (you should not, you default to Release https://github.com/monatis/clip.cpp/blob/1f140e027a3b9165a1edcdd4948bb44c3c2eacff/CMakeLists.txt#L7 , which sets `O3`)

monatis commented:
Yes `CMakeLists.txt` still requires some testing and optimizations (also for other arches like different arm versions) so it might be another PR when I take down my Raspberry PIs from the shelf. C++20 standard was just a guard but we can downgrade it now or after some tests. I can refactor the directory walk with the function from `common-clip` that does not require C++17. Then everything will be ok with C++11.

I'm ok with it as is, and then we can continue to further improve it. When #30 is merged, we can also the use batch inference when indexing images.

Green-Sky commented:
> Yes `CMakeLists.txt` still requires some testing and optimizations (also for other arches like different arm versions) so it might be another PR when I take down my Raspberry PIs from the shelf. C++20 standard was just a guard but we can downgrade it now or after some tests. I can refactor the directory walk with the function from `common-clip` that does not require C++17. Then everything will be ok with C++11.

yea, there have also been changes upstream (especially llama) , where you obviously ~copied~ based the code on.

> I'm ok with it as is, and then we can continue to further improve it. When #30 is merged, we can also the use batch inference when indexing images.

Yea. Let me run some small tests and then i will make it ready for review.

monatis commented:
> yea, there have also been changes upstream (especially llama) , where you obviously copied based the code on.

Exactly. The entire ggml-based projects are moving very fast, so I believe we don't need to target 100% production-grade multiplatform-ability. The speed of iterations and innovations is prioritized currently.

> Yea. Let me run some small tests and then i will make it ready for review.

Great. Thank you.

monatis commented:
USearch reports distance values instead of similarities, so I made changes to emphasize it. LGTM, merging. TODOs can be fixed later.

Thanks for your contributions.
